Lombiq.Analyzers.PowerShell.Tests/UnitTests/PowerShellAnalysisTests.cs

```diff
@@ -0,0 +1,70 @@
+﻿using CliWrap;
+using CliWrap.Buffered;
+using Lombiq.HelpfulLibraries.Cli.Helpers;
+using Shouldly;
+using System;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace Lombiq.Analyzers.PowerShell.Tests.UnitTests;
+
+public class PowerShellAnalysisTests
+{
+    private static readonly Command _powerShell = Cli.Wrap("pwsh");
+
+    private static readonly DirectoryInfo _testSolutions = new(Path.Combine("..", "..", "..", "..", "TestSolutions"));
+
+    [Fact]
+    public async Task DirectScriptInvocationShouldDisplayWarnings()
+    {
+        if (!await IsPowerShellCoreInstalledAsync()) return;
+
+        var result = await _powerShell
+            .WithWorkingDirectory(Path.GetFullPath(_testSolutions.FullName))
+            .WithArguments(new[] { "-c", "../Lombiq.Analyzers.PowerShell/Invoke-Analyzer.ps1 -IncludeTestSolutions" })
+            .WithValidation(CommandResultValidation.None)
+            .ExecuteBufferedAsync();
+
+        result.ExitCode.ShouldNotBe(0);
+        MessageShouldContainViolationCodes(result.StandardError);
+    }
+
+    [Theory]
+    [InlineData("Lombiq.Analyzers.PowerShell.PackageReference")]
+    [InlineData("Lombiq.Analyzers.PowerShell.ProjectReference")]
+    public async Task BuildShouldDisplayWarnings(string directory)
+    {
+        if (!await IsPowerShellCoreInstalledAsync()) return;
+
+        var solutionPath = _testSolutions.GetDirectories(directory).Single().FullName;
+
+        var exception = (InvalidOperationException)await Should.ThrowAsync(
+            () => DotnetBuildHelper.ExecuteStaticCodeAnalysisAsync(solutionPath),
+            typeof(InvalidOperationException));
+
+        MessageShouldContainViolationCodes(exception.Message);
```

Comment:
This passes even if I remove the content of Violate-Analyzers.ps1. Also add a positive test case.

Reply:
[Of course, becuase it's copied right back.](https://github.com/Lombiq/PowerShell-Analyzers/blob/4c113a69baa3c5b3fcc5530c0425632ffecabd0b/TestSolutions/Lombiq.Analyzers.PowerShell.PackageReference/Lombiq.Analyzers.PowerShell.PackageReference.csproj#L13-L15)
What is the utility of a positive test case? It's not clear to me why you'd want that.

Reply:
> I was editing the file under tools\Lombiq.Analyzers.PowerShell\TestSolutions but the same happens with the one under tools\Lombiq.Analyzers.PowerShell\TestSolutions\Lombiq.Analyzers.PowerShell.ProjectReference.

Works for me tho: [screen-recording.webm](https://user-images.githubusercontent.com/4007293/178357694-027d9504-c611-483f-874e-e39ba607c30d.webm)

> A correct script shouldn't cause analyzer violations. While we can't test every single case (just as with the negative test) we should at least attempt a smoke test and make sure that a known correct script indeed doesn't cause one (because we e.g. messed up the configuration, or the analyzer has an internal error).

Since we can't do it exhaustively I don't think it's a good idea to have positive tests. False sense of security. I've added check to the existing test to verify that it has exactly 4 violations, no more no less. Also added an extra line of code to the _Violate-Analyzers_ that will certainly trip if our settings file is not loaded.

Reply:
Now I see that the tests don't actually execute when `pwsh` is not available, and this was the issue for me. However, changing line 15 to use `"powershell"`, as it would be with an alias, all tests fail in every case. So this will fail in TC once the alias is present.

About the positive tests, looks good.

Reply:
Based on [this](https://teams.microsoft.com/l/message/19:2369a00506054c87b8b6ba37cc04270d@thread.skype/1657551388632?tenantId=959ede5a-38e9-44e6-adc8-85553beb8220&groupId=d6a5d775-6fb9-41dc-a7bb-481d4f884046&parentMessageId=1657538205170&teamName=Lombiq%20Developers&channelName=Infrastructure%20team&createdTime=1657551388632) it looks like we will need actual pwsh on TC anyway. Also that's better becuase then the script will be able to update its own module version as user without need for admin intervention when future versions of PSScriptAnalyzer come out.

Reply:
If I replace the string on line 15 with "powershell", then the tests still pass for me:
![image](https://user-images.githubusercontent.com/4007293/178373659-93362851-eff0-4db7-8a99-bb4395f7e180.png)
I have already tested the script on Windows PowerShell in Utility Scripts so I know it's good on both my machine and GH's windows-latest VMs.

Don't forget that you have to install the module from within an admin session of Windows PowerShell first. Otherwise the tests _should_ fail. I have uninstalled it and then executing DirectScriptInvocationShouldDisplayWarnings results in this error while targetting `powershell`:
![image](https://user-images.githubusercontent.com/4007293/178375203-7763865c-adf0-4c5a-8a98-d8bd031ff1ad.png)

Reply:
Yes, the module is installed.

I should've copied the errors here, obviously... Yesterday I got errors related to Invoke-Analyzers not being found when it was actually there. Now with the latest code not all tests but just the two `BuildShouldDisplayWarnings` fail with the following (same for the other):

```
  Message: 
Shouldly.ShouldAssertException : exception.Message
    should match
"The command [^\n]+Invoke-Analyzer.ps1 -ForMsBuild -IncludeTestSolutions[^\n]+exited with code 4\."
    but was
"The dotnet.exe build ..\..\..\..\TestSolutions\Lombiq.Analyzers.PowerShell.ProjectReference\Lombiq.Analyzers.PowerShell.ProjectReference.sln --no-incremental --nologo --warnaserror --consoleLoggerParameters:NoSummary --verbosity:quiet -p:TreatWarningsAsErrors=true -p:RunAnalyzersDuringBuild=true command failed with the output below.
E:\ShortPath\OSOCE3\tools\Lombiq.Analyzers.PowerShell\TestSolutions\Lombiq.Analyzers.PowerShell.ProjectReference\Violate-Analyzers.ps1(1,10): error PSUseSingularNouns: The cmdlet 'Violate-Analyzers' uses a plural noun. A singular noun should be used instead. [E:\ShortPath\OSOCE3\tools\Lombiq.Analyzers.PowerShell\TestSolutions\Lombiq.Analyzers.PowerShell.ProjectReference\Lombiq.Analyzers.PowerShell.ProjectReference.csproj]
E:\ShortPath\OSOCE3\tools\Lombiq.Analyzers.PowerShell\TestSolutions\Lombiq.Analyzers.PowerShell.ProjectReference\Violate-Analyzers.ps1(6,7): error PSAvoidUsingCmdletAliases: 'echo' is an alias of 'Write-Output'. Alias can introduce possible problems and make scripts hard to maintain. Please consider changing alias to its full content. [E:\ShortPath\OSOCE3\tools\Lombiq.Analyzers.PowerShell\TestSolutions\Lombiq.Analyzers.PowerShell.ProjectReference\Lombiq.Analyzers.PowerShell.ProjectReference.csproj]
E:\ShortPath\OSOCE3\tools\Lombiq.Analyzers.PowerShell\TestSolutions\Lombiq.Analyzers.PowerShell.ProjectReference\Violate-Analyzers.ps1(1,10): error PSUseApprovedVerbs: The cmdlet 'Violate-Analyzers' uses an unapproved verb. [E:\ShortPath\OSOCE3\tools\Lombiq.Analyzers.PowerShell\TestSolutions\Lombiq.Analyzers.PowerShell.ProjectReference\Lombiq.Analyzers.PowerShell.ProjectReference.csproj]
E:\ShortPath\OSOCE3\tools\Lombiq.Analyzers.PowerShell\TestSolutions\Lombiq.Analyzers.PowerShell.ProjectReference\Violate-Analyzers.ps1(6,32): error PSAvoidUsingEmptyCatchBlock: Empty catch block is used. Please use Write-Error or throw statements in catch blocks. [E:\ShortPath\OSOCE3\tools\Lombiq.Analyzers.PowerShell\TestSolutions\Lombiq.Analyzers.PowerShell.ProjectReference\Lombiq.Analyzers.PowerShell.ProjectReference.csproj]
E:\ShortPath\OSOCE3\tools\Lombiq.Analyzers.PowerShell\Lombiq.Analyzers.PowerShell\Lombiq.Analyzers.PowerShell.targets(40,5): error MSB3073: The command "powershell E:\ShortPath\OSOCE3\tools\Lombiq.Analyzers.PowerShell\Lombiq.Analyzers.PowerShell\Invoke-Analyzer.ps1 -ForMsBuild -IncludeTestSolutions" exited with code 1. [E:\ShortPath\OSOCE3\tools\Lombiq.Analyzers.PowerShell\TestSolutions\Lombiq.Analyzers.PowerShell.ProjectReference\Lombiq.Analyzers.PowerShell.ProjectReference.csproj]
"

Additional Info:
    The Invoke-Analyzer script should've returned 4 exit code because that's the exact number of expected violations.

  Stack Trace: 
PowerShellAnalysisTests.BuildShouldDisplayWarnings(String directory) line 48
ExceptionDispatchInfo.Throw()
TaskAwaiter.ThrowForNonSuccess(Task task)
TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
TaskAwaiter.GetResult()
--- End of stack trace from previous location ---
ExceptionDispatchInfo.Throw()
TaskAwaiter.ThrowForNonSuccess(Task task)
TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
TaskAwaiter.GetResult()
ExceptionDispatchInfo.Throw()
TaskAwaiter.ThrowForNonSuccess(Task task)
TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
TaskAwaiter.GetResult()
```

Reply:
I think the problem was that you can do `pwsh Invoke-Analyzer.ps1` but you can't do `powershell Invoke-Analyzer.ps1`. It has to be `powershell -File Invoke-Analyzer.ps1`. I've updated `<Exec>` in the targets file accordingly.

Besides that I'm unsure if Windows PowerShell handles exit codes as rigorously as you'd expect from a multiplatform shell becuase Windows traditionally didn't give a damn about them. Of course the tests specifically use `pwsh` and there is no good reason not to have it installed.

Reply:
Now it works, thank you.